File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		Stack<Integer> stack = new Stack<Integer>();
		
		
		int N = sc.nextInt();
		int[] arr = new int[N];
		
		for(int i=0;i<N;i++) {
			arr[i] = sc.nextInt();
		}
		sc.close();
		
		int now_num = 1; //들어가야할 숫자
		int j=0; //줄서있는 순서 0~N-1
		while(now_num<=N) {
			
			//원래 열 다 끝날 경우
			if(j==N) {
//				System.out.println(stack);
				allstack(stack, now_num);
				return;
			}
			
			//원래 열에서 넣을 수 있는지
			if(arr[j]==now_num) {
//				System.out.println("원래 열에서" + now_num + "넣음");
				now_num++;
				j++;
			}
			//스택의 꼭대기에서 넣을 수 있는지
			else if(!stack.isEmpty()&&stack.peek()==now_num) {
//				System.out.println("pop 실행으로" + now_num + "넣음");
				stack.pop();
				now_num++;
			}
			//못넣으면 스택에 저장
			else {
				stack.push(arr[j]);
				j++;
			}
		}
		System.out.println("Nice");
		
		// 10
		// 5 4 3 2 1 6 7 8 9 10
		
		// 2 1 3 인경우 바로 들어갈 수 있기 때문에 Nice다.

	}
	
	public static void allstack(Stack<Integer> stack, int now_num) {
		while(!stack.isEmpty()) {
			if(stack.peek()==now_num) {
				stack.pop();
				now_num++;
			}
			else {
				System.out.println("Sad");
				return;
			}
		}
		System.out.println("Nice");
	}
	

}

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.cc
#include <bits/stdc++.h>
using namespace std;

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int N, arr[1001];
    cin >> N;
    for (int i=1; i<=N; i++){
        cin >> arr[i];
    }

    stack<int> wait;
    wait.push(-1);  // 빈 스택 에러방지

    int now = 1;

    for (int i=1; i<=N; i++){

        while(wait.top()!=i){
            wait.push(arr[now]);
            now++;
            if (now==N+2){
                cout << "Sad";
                return 0;
            }
        }
        wait.pop();
    }

    // 위의 for문을 돌렸는데 스택에 -1 외에 더 남아있다면 못준거임
    wait.pop();  // -1 삭제
    if (wait.empty()) cout << "Nice";
    else cout << "Sad";
    
    return 0;
}